src/agents/memory/providers/redis.py

```diff
@@ -0,0 +1,245 @@
+from __future__ import annotations
+
+import json
+import time
+from typing import TYPE_CHECKING
+
+from agents.memory.session import SessionABC
+
+try:
+    import redis.asyncio as redis
+
+    if TYPE_CHECKING:
+        from agents.items import TResponseInputItem
+
+except ImportError as err:
+    raise ImportError("redis and openai-agents packages are required") from err
+
+
+class RedisSession(SessionABC):
+    """Redis-based implementation of session storage.
+
+    This implementation stores conversation history in Redis using lists and hashes.
+    Each session uses a Redis list to store messages in chronological order and
+    a hash to store session metadata.
+    """
+
+    def __init__(
+        self,
+        session_id: str,
+        redis_url: str = "redis://localhost:6379",
+        db: int = 0,
+        session_prefix: str = "agent_session",
+        messages_prefix: str = "agent_messages",
+        ttl: int | None = None,
+    ):
+        """Initialize the Redis session.
+
+        Args:
+            session_id: Unique identifier for the conversation session
+            redis_url: Redis connection URL. Defaults to 'redis://localhost:6379'
+            db: Redis database name. Defaults to `default`
+            session_prefix: Prefix for session metadata keys. Defaults to 'agent_session'
+            messages_prefix: Prefix for message list keys. Defaults to 'agent_messages'
+            ttl: Time-to-live for session data in seconds. If None, data persists indefinitely
+        """
+        self.session_id = session_id
+        self.redis_url = redis_url
+        self.db = db
+        self.session_prefix = session_prefix
+        self.messages_prefix = messages_prefix
+        self.ttl = ttl
+
+        # Redis keys for this session
+        self.session_key = f"{session_prefix}:{session_id}"
+        self.messages_key = f"{messages_prefix}:{session_id}"
+
+        self._redis_client: redis.Redis | None = None
+
+    async def _get_redis_client(self) -> redis.Redis:
+        """Get or create Redis client."""
+        if self._redis_client is None:
+            self._redis_client = redis.from_url(
+                self.redis_url,
+                db=self.db,
+                decode_responses=True,
+                retry_on_error=[redis.BusyLoadingError, redis.ConnectionError],
+                retry_on_timeout=True,
+            )
+        return self._redis_client
+
+    async def _ensure_session_exists(self, client: redis.Redis) -> None:
+        """Ensure session metadata exists in Redis."""
+        current_time = time.time()  # Use float for higher precision
+
+        # Check if session already exists
+        exists = await client.exists(self.session_key)
```

Comment:
Is there a possible race here if 2 clients tried to create the same session at once? Does the `client.exist` and `client.hset` need to be in a multi/exec/watch?